packages/astro/src/astro-components/control/ProtectCSR.astro

```diff
@@ -0,0 +1,52 @@
+---
+import type { ProtectProps } from '../../types';
+
+type Props = Omit<ProtectProps, 'condition'>;
+
+const { role, permission } = Astro.props;
+---
+
+<clerk-protect
+  data-role={role}
+  data-permission={permission}
+>
+  <div data-clerk-control-slot-default>
+    <slot />
+  </div>
+  <div data-clerk-control-slot-fallback>
+    <slot name="fallback" />
+  </div>
+</clerk-protect>
+
+<script>
+import { $sessionStore } from '@clerk/astro/client';
+import { type AuthState, BaseClerkControlElement } from './BaseClerkControlElement';
+import type { CheckAuthorization } from '@clerk/types';
+
+class ClerkProtect extends BaseClerkControlElement {
+  private defaultSlot: HTMLDivElement | null = null;
+  private fallbackSlot: HTMLDivElement | null = null;
+
+  protected onAuthStateChange(state: AuthState): void {
+    this.defaultSlot = this.querySelector('[data-clerk-control-slot-default]');
+    this.fallbackSlot = this.querySelector('[data-clerk-control-slot-fallback]');
+
+    const has = $sessionStore.get()?.checkAuthorization;
+
+    const role = this.dataset.role;
+    const permission = this.dataset.permission;
+    const isUnauthorized =
+      !state.userId || ((role || permission) && !has?.({ role, permission } as Parameters<CheckAuthorization>[0]));
+
+    if (this.defaultSlot) {
+      isUnauthorized ? this.defaultSlot.setAttribute('hidden', '') : this.defaultSlot.removeAttribute('hidden');
```

Comment:
For the `hidden` attribute I wondered if it's best practice in terms of accessibility. According to https://www.a11yproject.com/posts/how-to-hide-content/ it should be fine:

> Another way to hide content both visually and from assistive technology is the [HTML5 hidden attribute](https://html.spec.whatwg.org/multipage/interaction.html#the-hidden-attribute).

However, both that resource and https://meowni.ca/hidden.is.a.lie.html mention that this might possibly be overridden by other CSS:

> to increase the specificity of the hidden attribute, you can add the following snippet to your CSS:
>
> This is unfortunately a lie: the hidden rule is a User Agent style, which means it's less specific than a moderate sneeze [[ref]](https://bugs.chromium.org/p/chromium/issues/detail?id=566168#c10). This means that your favourite display style will override it

So it might be worth it to toggle this inline style, too:

```css
[hidden] { display: none !important}
```

Reply:
Good read, updated!